cranelift-sizeless-vector.md

```diff
@@ -0,0 +1,112 @@
+//! SP at function entry ----->  | return address            |
+//!                              +---------------------------+
+//!                              |          ...              |
+//!                              | clobbered callee-saves    |
+//! unwind-frame base     ---->  | (pushed by prologue)      |
+//!                              +---------------------------+
+//! FP after prologue -------->  | FP (pushed by prologue)   |
+//!                              +---------------------------+
+//!                              | sizeless stack slots      |
+//!                              | (accessed via FP)         |
+//!                              |          ...              |
+//!                              +---------------------------+
+//!                              | spill slots               |
+//!                              | (accessed via nominal SP) |
+//!                              |          ...              |
+//!                              | stack slots               |
+//!                              | (accessed via nominal SP) |
+//! nominal SP --------------->  | (alloc'd by prologue)     |
+//! (SP at end of prologue)      +---------------------------+
+//!                              | [alignment as needed]     |
+//!                              |          ...              |
+//!                              | args for call             |
+//! SP before making a call -->  | (pushed at callsite)      |
+//!                              +---------------------------+
+//!
+//!   (low address)
+//! ```
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main design decision is to have a vector type that isn't comparable to the existing vector types, which means that no existing paths can accidently try to treat them as such. Although setting the minimum size of the opaque type to 128 bits still allows us to infer the minimum number of lanes for a given lane type. The flexible vector specification also provides specific operations for lane accesses and shuffles so these aren't operations that we have to handle with our existing operations.
+
+It's possible that the hardware vector length will be fixed, so one alternative would be to generate IR with fixed widths using information from the backend. The one advantage is that we'd not have to add sizeless types in the IR at all. But there are two main disadvantages with this approach:
+- In an ahead-of-time setting, Cranelift would not be able to take advantage of larger vectors where the width is implementation defined. It would be possible to target architecture extensions such as Intel's AVX2 and AVX-512, which have static sizes, but not for architectures like Arm's SVE.
+- It is also currently undecided whether the flexible vector specification will include operations to set the vector length during program execution, so we shouldn't design out this possibility.
```

Comment:
@cfallin, @sparker-arm: I just caught up on all of the recent discussion above for how to implement this proposal in a platform-agnostic way, etc. It's good stuff and I think I understand a lot better your thoughts from the examples given.

I do think, though, that this point is unaddressed by the comments about creating dynamic types in the prologue: if we create a dynamic type of a certain length, how then could we change that length dynamically? @penzn included a [set of `set_length` instructions](https://github.com/WebAssembly/flexible-vectors/blob/main/proposals/flexible-vectors/FlexibleVectorsThirdTier.md) (with caveats about needing more discussion). @sunfishcode had an alternate design idea here and, @sparker-arm, you've commented in the issue (https://github.com/WebAssembly/flexible-vectors/issues/21).

My point is: I sense a mismatch with this dynamic type idea and `set_length` and I think we should figure that out (here or in the flexible vectors repository) before something gets designed that precludes it.

Reply:
Ah, that is a really interesting twist!

I haven't fully digested the examples in that issue but e.g. the existence of a `vec_loop` primitive together with dynamically-sized vectors makes me think that somehow there should be a lowering step involved: once the code reaches CLIF, there is no more vector-specific loop primitive, so whatever blocking or grouping that the `vec_loop` does, and how it interacts with the truly dynamic and resizable size, should be lowered onto "fixed for the function invocation"-sized chunks.

Also from a codegen perspective, resizable types have implications on stackframe layout: what does it mean for the regalloc if we end up having to spill `v0`, but `v0` has a type whose size changes throughout the function body?

So I think there is a difficulty/complexity jump from "dynamically sized, known at function entry" and "dynamically sized, changes size at arbitrary times" and we should figure out if we can lower the latter (at Wasm level) onto the former (at CLIF level) if we can. Curious if there are other ideas here though!

Reply:
> resizable types have implications on stackframe layout

Brainstorming: it feels like there are two parts here, 1) the size of the slot that needs to be spilled, which could be the "max vector size on this uarch," like the `dt0` example above, and 2) the number of lanes used by that vector, which could be modified by `set_length`. Even if only the lowest few lanes are used, the entire vector can be spilled, right?

Reply:
I guess I'm saying that the type need not be resizable, but there should be some other way to express how much of that type is being used.

Reply:
Yep, that's true; maybe what we want is an extra parameter on the operations that actually use the vector, to indicate how many lanes to operate on. So the type is `Nxi32` where N is the max size (could be dynamic, but fixed at function entry); and we then do an add on `Mxi32` units, where `M` was set by the latest `set_length`. I think this would imply tracking a separate "current length" variable for each dynamically-typed-and-resizable vector in the Wasm translator when building the SSA for CLIF. Does that seem reasonable?

Reply:
I think Petr's idea for `set_length` was even more dynamic than that--it can accept arbitrary `i32` values from the stack. I might be misunderstanding what you mean, but we wouldn't know what that `M` is until runtime, so it couldn't be fixed during translation. But maybe you mean that that "variable" in translation would be made available for `set_length` to change at runtime?

Reply:
Yes, exactly, it would be a variable that has an SSA `Value` at any given program point, like other Wasm locals, so truly dynamic. Then we could have something like `add_dynamic.dt0 v0, v1, v2` where `v0` and `v1` have type `dt0` (which is `Nxi32`, set to the max possible size), and `v2` is an `i32` that tells the op how many lanes to operate on.

Reply:
Just to note, my understanding was the vec_loop construct won't be pursued and instead predication, as @cfallin is suggesting here, will be the way to go. We can add predicated operations for those if/when it happens. As @abrown said, when it comes to spilling, I also presume predication (set_length) won't affect this, we'll still spill a whole register.